apis/python/src/tiledbsoma/io/registration/ambient_label_mappings.py

```diff
@@ -0,0 +1,405 @@
+
+import tiledbsoma
+import tiledbsoma.logging
+
+from .id_mappings import AxisIDMapping, ExperimentIDMapping, get_dataframe_values
+
+
+@attrs.define(kw_only=True)
+class AxisAmbientLabelMapping:
+    """
+    For all the to-be-appended AnnData/H5AD inputs in SOMA multi-file append-mode ingestion, this
+    class tracks the mapping of input-data ``obs`` or ``var`` ID-column name (barcode ID, gene
+    symbol) to SOMA join IDs for SOMA experiment ``obs`` or ``var``.
+
+    See module-level comments for more information.
+    """
+
+    data: Dict[str, int]
+    field_name: str
+
+    def get_next_start_soma_joinid(self) -> int:
+        """Once some number of input files have been registered for an ``obs`` or ``var``
+        axis, this returned the next as-yet-unused SOMA join ID for the axis."""
+        if len(self.data) == 0:
+            return 0
+        else:
+            return max(self.data.values()) + 1
+
+    def id_mapping_from_values(self, input_ids: Sequence[Any]) -> AxisIDMapping:
+        """Given registered label-to-SOMA-join-ID mappings for all registered input files for an
+        ``obs`` or ``var`` axis, and a list of input-file 0-up offsets, this returns an int-to-int
+        mapping from a single input file's ``obs`` or ``var`` axis to the registered SOMA join IDs.
+        """
+        soma_joinids = []
+        for i, input_id in enumerate(input_ids):
+            if input_id not in self.data:
+                raise ValueError(f"input_id {input_id} not found in registration data")
+            soma_joinids.append(self.data[input_id])
+        return AxisIDMapping(data=soma_joinids)
+
+    def id_mapping_from_dataframe(self, df: pd.DataFrame) -> AxisIDMapping:
+        """Given registered label-to-SOMA-join-ID mappings for all registered input files for an
+        ``obs`` or ``var`` axis, and an input file's dataframe with its 0-up offsets, this returns
+        an int-to-int mapping from a single input file's ``obs`` or ``var`` axis to the registered
+        SOMA join IDs.
+        """
+        values = get_dataframe_values(df, self.field_name)
+        return self.id_mapping_from_values(values)
+
+    @classmethod
+    def from_isolated_dataframe(
+        cls,
+        df: pd.DataFrame,
+        *,
+        index_field_name: Optional[str] = None,
+    ) -> Self:
+        """Factory method to compute an axis label-to-SOMA-join-ID mapping for a single dataframe in
+        isolation. This is used when a user is ingesting a single AnnData/H5AD to a single SOMA
+        experiment, not in append mode, but allowing us to still have the bulk of the ingestor code
+        to be non-duplicated between non-append mode and append mode.
+        """
+        tiledbsoma.logging.logger.info("Registration: registering AnnData dataframe.")
+
+        if index_field_name is None:
+            index_field_name = df.index.name
+        if index_field_name is None:
+            index_field_name = "index"
```

Comment:
suggestion: use `or`s to shorten this:

`index_field_name = index_field_name or df.index.name or "index"`